agentrs/src/listener.rs

```diff
@@ -0,0 +1,120 @@
+use crate::tasks::*;
+use crate::ws::client::WebSocket;
+use std::time::Duration;
+use tokio::runtime::Runtime;
+use tracing::{error, info};
+
+use tokio::runtime;
+
+pub struct Service {
+    state: State,
+}
+
+async fn build_tasks() -> anyhow::Result<Tasks, anyhow::Error> {
+    let mut tasks = Tasks::new();
+
+    let ws = WebSocket::new()?;
+
+    tasks.register(ws);
+    Ok(tasks)
+}
+
+enum State {
+    Stopped,
+    Running {
+        shutdown_handle: ShutdownHandle,
+        runtime: Runtime,
+    },
+}
+
+impl Service {
+    pub fn new() -> Self {
+        Self {
+            state: State::Stopped,
+        }
+    }
+
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        let runtime = runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .expect("failed to create runtime");
+
+        let tasks = runtime.block_on(build_tasks())?;
+
+        let mut join_all = futures::future::select_all(
+            tasks.inner.into_iter().map(|child| Box::pin(child.join())),
+        );
+
+        runtime.spawn(async {
+            loop {
+                let (result, _, rest) = join_all.await;
+                match result {
+                    Ok(Ok(())) => {
+                        println!("A task terminated gracefully")
+                    }
+                    Ok(Err(error)) => error!("A task failed {:?}", error),
+                    Err(error) => error!("Something went very wrong with a task {:?}", error),
+                }
+
+                if rest.is_empty() {
+                    break;
+                } else {
+                    join_all = futures::future::select_all(rest);
+                }
+            }
+        });
```

Comment:
Try something like this:
```suggestion
        let mut join_set = JoinSet::new();
        for child in tasks.inner {
            join_set.spawn(child.join());
        }

        runtime.spawn(async move {
            while let Some(result) = join_set.join_next().await {
                match result {
                    Ok(Ok(_)) => {
                        println!("A task terminated gracefully");
                    }
                    Ok(Err(error)) => {
                        error!("A task failed {:?}", error);
                    }
                    Err(error) => {
                        error!("Something went very wrong with a task {:?}", error);
                    }
                }
            }
        });```

Reply:
Adding join set will need to change the shutdown 
i think will need to coordinate the shutdown with the join set
everything inside a for loop 
```
for .. max_count {
select {
   maybe = join_set.join_next().. 
     your match...

   all_closed
   time to tick // sleep 
}
}
```

Reply:
JoinSet is not Sync; it should be owned and polled by a single task—the one coordinating shutdown.

Reply:
i might wrong, could be I could change it easily